src/antpm-downloader.1
.\" -*- coding: utf-8-unix -*-
.\" groff -man -Tutf8 antpm-downloader.1
.\" Copyright © 2013 by Kristóf Ralovich
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\"
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\" License.
.\"
.TH antpm-downloader 1 "2013-10" "antpm-downloader v1.9" "ANT-FS client"
.SH "NAME"
antpm-downloader \- console based ANT-FS information retrieval client for Garmin GPS products.
.SH "SYNOPSIS"
.BI "antpm-downloader [ " "OPTIONS" " ] "
.SH DESCRIPTION
.B "antpm-downloader"
is a small command-line utility to retrieve information (sports
tracks, etc.) from Garmin GPS products over the wireless ANT-FS
protocols. Currently supported devices are the Forerunner 310XT. It is
expected that the Forerunner 60, Forerunner 610, Forerunner 910XT,
FR70 and Swim models work, but user feedback is required.
.SH "OPTIONS"
.TP
\fB\-P\fR [ \fB\-\-pairing\fR ]
Force pairing first
.TP
\fB\-\-dir\-only\fR
Download and list device directory
.HP
\fB\-D\fR [ \fB\-\-download\fR ] arg Download a single file (hex id e.g. 0x12FB) from device
.TP
\fB\-\-erase\fR arg
Erase a single file (hex id e.g. 0x12FB) from device
.TP
\fB\-\-v\fR arg
Adjust verbosity level, varies in [1, 6]
.SS "Help options"
.TP
\fB\-h\fR [ \fB\-\-help\fR ]
produce help message
.TP
\fB\-V\fR [ \fB\-\-version\fR ]
Print version information
.SH LINUX USER'S GUIDE
.PP
Make sure you're using the latest released version of ANT+minus\.
.PP
Make sure that the cp210x kernel module is loaded: "sudo modprobe cp210x"\. Verify by looking at non-empty output of "lsmod | grep cp210x"\.
.PP
Make sure that the USB2ANT stick is connected to the machine\.
.PP
Make sure cp210x driver was assigned to your USB2ANT\. Verify by "dmesg | grep tty" and look for a line similar to: "[ 3274.854726] usb 6-2: cp210x converter now attached to ttyUSB0". Also make sure "ls /dev/ttyUSB*" lists your device node.
.PP
Your user must be able to access the /dev/ttyUSB0 device. This can be achieved e.g. by the installing the
.UR https://github.com/ralovich/antpm/blob/master/scripts/80-ant-stick.rules
.BR https://github.com/ralovich/antpm/blob/master/scripts/80-ant-stick.rules
.UE
udev rule into \fI/etc/udev/rules.d\fP. Restart udev. Verify by "cat /dev/ttyUSB0".
.PP
Install the antpm binary package.
.PP
Enable "pairing mode" on your watch.
.PP
Bring your watch to ~50cm from the USB2ANT stick. Sometimes avoiding the proximity of a WiFi network helps.
.PP
Run
.B "antpm-downloader"
on your machine. This will attempt connecting to the watch and pair with it first.
.PP
During pairing, you have to click "Yes" on the watch to complete pairing.
.PP
Afterwards all activity information from the watch is downloaded to your machine. This can take some time.
.PP
The downloaded files are stored on your machine to the folder ConfigFolder/*device-id*/*current-date-time* (where ConfigFolder equals $XDG_CONFIG_HOME which usually points to $HOME/.config/antpm/ or $ANTPM_DIR/ if that is set).
.PP
The 310XT stores all information in .fit files. These files can describe waypoints/tracks/activities.
.PP
To convert the downloaded .fit files to a more commonly used .gpx format, issue the command "antpm-fit2gpx -F ConfigFolder/*device-id*/*current-date-time*" e.g. "antpm-fit2gpx -F ~/.config/antpm/4531082473/2013_03_26_10_58_19". This will populate the given folder with .gpx files.
.SH HOW TO SUBMIT A BUG-REPORT
.PP
Please open a bug report at
.UR "https://github.com/ralovich/antpm"
.BR "https://github.com/ralovich/antpm"
.UE
.PP
In your report please include:
.LP
.RS
The log of the program in question (usually saved under ~/.config/antpm)?
.RE
.LP
.LP
.RS
What steps will reproduce the problem?
.RE
.LP
.LP
.RS
What is the expected output? What do you see instead?
.RE
.LP
.LP
.RS
What version of ANT+minus are you using? On what operating system? Is it pre-built of from sources?
.RE
.LP
.LP
.RS
What GPS are you trying to use? What firmware version?
.RE
.LP
.SH WINDOWS USER'S GUIDE
.PP
Make sure that the USB2ANT stick is connected to the machine.
.PP
Install the libusb driver by running "3rd_party/libusb-win32-bin-1.2.6.0/driver_installer/inf-wizard.exe"
.SH "DEVICES"
Wireless ANT communications happens over a USB2ANT dongle. Under Linux
the dongle is driven by
.B cp210x
kernel module, and presented to user-space as a USB serial port,
.B /dev/ttyUSBxxx
where
.B xxx
is an integer. The dongle is typically carrying the USB vendor id/product id
pair of 0fcf/1004 or 0fcf/1008. Recent Linux systems employ
.B udev
to handle device nodes and permissions, thus appropriate
.SH "FILES"
.TP
\fB/lib/modules/`uname \-r`/kernel/drivers/usb/serial/cp210x.ko\fR
Kernel driver module for USB2ANT dongle.
.TP
.B /dev/ttyUSBxxx
Serial device created for dongle.
.TP
.B $XDG_CONFIG_HOME/antpm ($HOME/.config/antpm)
Default config folder (downloaded content is also stored here).
.TP
.B ANTPM_DIR
The
.B ANTPM_DIR
environment variable is honored to override the default config folder.
.SH "AUTHOR"
.B antpm-downloader
was written by Kristóf Ralovich. This man page is Copyright ©
2010-2013 Kristóf Ralovich.
.SH "SEE ALSO"
.UR "https://github.com/ralovich/antpm"
.BR "https://github.com/ralovich/antpm"
.UE


debian/changelog
antpm (1.8-1) UNRELEASED; urgency=low

  * GarminConvert: add a method that offsets the timestamps
  * AntMessage: bugfix: correctly convert file creation time

 -- RALOVICH, Kristóf <tade60@freemail.hu>  Sat, 06 Jul 2013 00:46:18 +0200

antpm (1.7-1) UNRELEASED; urgency=low

  * antpm-*: recongize -h option for --help too
  * fit: this is a run-time warning
  * docs: start writing user's guide
  * docs: how to submit a bug report
  * docs: add a note on installing the udev rule for user access to the
    USB2ANT stick
  * docs: refine udev procedure
  * docs: split maintainer documentation from end-user documentation
  * FIT: bugfix: don't offset timestamps twice
  * bugfix: this was a crippling bug
  * docs: add todo
  * new version

 -- RALOVICH, Kristóf <tade60@freemail.hu>  Sat, 06 Jul 2013 00:11:34 +0200

antpm (1.6-3) UNRELEASED; urgency=low

  [ RALOVICH, Kristóf ]
  * docs: more notes
  * this change was only necessary for win64
  * debian: refine c++ compiler dependency
  * serial: add more verbosity to output

  [ Ralovich Kristof ]

 -- Ralovich Kristof <tade@cg.iit.bme.hu>  Wed, 26 Jun 2013 11:43:35 +0200

antpm (1.6-2) UNRELEASED; urgency=low

  [ Kristof Ralovich ]
  * script to generate changelog from git

  [ Ralovich Kristof ]

 -- Ralovich Kristof <tade@cg.iit.bme.hu>  Tue, 25 Jun 2013 17:10:28 +0200

antpm (1.6-0) UNRELEASED; urgency=low

  [ RALOVICH, Kristóf ]
  * include Release builds in a .deb. package
  * clarify build- and run-time dependencies
  * add todo
  * fix build with clang 3.0
  * add some documentation
  * fix build warning
  * git should ignore build folder
  * AntFr310XT2::createDownloadFolder()
  * further define folder name based on device S/N
  * cleanup
  * deb: lower build time dependencies
  * create download folder after knowing the device id
  * fit2gpx: option to decode ANT root folder
  * initial, file based logging
  * docs: note on performance
  * fix output
  * use centralized logging in more places
  * fix output
  * introduce namespace antpm
  * namespace conversion
  * missing namespace
  * more unified logging
  * more unified logging
  * cleanup
  * fix an assertion failure
  * unified logging replaced loggerc()
  * reindent
  * define NDEBUG in release
  * avoid crashes

  [ unknown ]
  * build fixes

  [ RALOVICH, Kristóf ]
  * add todo
  * add todo
  * move GARMIN_EPOCH to antdefs.hpp
  * bugfix: fix thinko
  * try setting date of .fit files
  * change logging
  * first stub for centralized per-device setting

  [ RALOVICH, Kristof ]
  * missing include
  * cmake: fix linking under win
  * previously the x86 dll hada weird name...
  * cmake: make sure the correct generator is invoked
  * fix 64-bit linking
  * cmake: fix determination of win64
  * scripts: build helper for win64
  * antpm-downloader: bugfix
  * update discriution list

  [ Kristof Ralovich ]
  * deb: relax build time dependencies

  [ RALOVICH, Kristof ]
  * deb: libusb was missing as a build-dep

  [ Kristof Ralovich ]
  * bugfix: resolve gcc-4.4 build issues

  [ RALOVICH, Kristof ]
  * pkgconfig should look for libusb-1.0

  [ Kristof Ralovich ]
  * cleanup
  * fix missing include
  * debian: fixing dependencies
  * account for boost:filesystem version changes

  [ RALOVICH, Kristof ]
  * cmake: fix finding libusb under nix

  [ Ralovich Kristof ]

 -- Ralovich Kristof <tade@cg.iit.bme.hu>  Tue, 25 Jun 2013 17:02:17 +0200

antpm (1.0-1) UNRELEASED; urgency=low

  * Initial release.

 -- Kristof Ralovich <tade60@freemail.hu>  Tue, 21 Mar 2013 16:22:10 +0100


src/antpm-downloader.cpp
// -*- mode: c++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2; coding: utf-8-unix -*-
// ***** BEGIN LICENSE BLOCK *****
//////////////////////////////////////////////////////////////////////////
// Copyright (c) 2011-2014 RALOVICH, Kristóf                            //
//                                                                      //
// This program is free software; you can redistribute it and/or modify //
// it under the terms of the GNU General Public License as published by //
// the Free Software Foundation; either version 3 of the License, or    //
// (at your option) any later version.                                  //
//                                                                      //
// This program is distributed in the hope that it will be useful,      //
// but WITHOUT ANY WARRANTY; without even the implied warranty of       //
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        //
// GNU General Public License for more details.                         //
//                                                                      //
//////////////////////////////////////////////////////////////////////////
// ***** END LICENSE BLOCK *****

#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <vector>

#include "AntMessage.hpp"
#include "AntFr310XT.hpp"
#include "common.hpp"
#include "Log.hpp"

#include <boost/program_options.hpp>


#ifdef _WIN32
# include <Windows.h>
#elif defined(__linux)
# include <signal.h>
#endif

namespace po = boost::program_options;
using namespace std;
using namespace antpm;


namespace antpm
{

template<>
std::unique_ptr<Log>
ClassInstantiator<Log>::make_unique()
{
  mkDirNoLog(getConfigFolder().c_str());
  std::string l=getConfigFolder() + "/antpm_" + getDateString() + ".txt";
  return std::make_unique<Log>(l.c_str());
}

}



boost::function<void(void)> stopFunc;

static
void
stopIt()
{
  if(stopFunc)
    stopFunc();
}

#ifdef _WIN32
BOOL CtrlHandler( DWORD fdwCtrlType )
{
  switch( fdwCtrlType )
  {
    // Handle the CTRL-C signal.
    case CTRL_C_EVENT:
      printf( "Ctrl-C event\n\n" );
      //Beep( 750, 300 );
      stopIt();
      return( TRUE );

      // CTRL-CLOSE: confirm that the user wants to exit.
    case CTRL_CLOSE_EVENT:
      //Beep( 600, 200 );
      printf( "Ctrl-Close event\n\n" );
      stopIt();
      return( TRUE );

      // Pass other signals to the next handler.
    case CTRL_BREAK_EVENT:
      //Beep( 900, 200 );
      printf( "Ctrl-Break event\n\n" );
      stopIt();
      return FALSE;

    case CTRL_LOGOFF_EVENT:
      //Beep( 1000, 200 );
      printf( "Ctrl-Logoff event\n\n" );
      stopIt();
      return FALSE;

    case CTRL_SHUTDOWN_EVENT:
      //Beep( 750, 500 );
      printf( "Ctrl-Shutdown event\n\n" );
      stopIt();
      return FALSE;

    default:
      return FALSE;
  }
}
#elif (defined(__linux) || defined(__APPLE__) || defined(__GNU__))
void my_handler(int s)
{
  printf("Caught signal %d\nantpm teardown initiated\n\n",s);
  stopIt();
}
#endif


int
main(int argc, char** argv)
{
  printf("press Ctrl-C to exit\n");
#ifdef _WIN32
  if( !SetConsoleCtrlHandler( (PHANDLER_ROUTINE) CtrlHandler, TRUE ) )
  {
    printf( "\nERROR: Could not set control handler...");
  }
#elif (defined(__linux) || defined(__APPLE__) || defined(__GNU__))
  signal(SIGINT, my_handler);
#endif

  antpm::Log::instance()->addSink(std::cout);
#ifdef NDEBUG
  antpm::Log::instance()->setLogReportingLevel(antpm::LOG_INF);
#else
  antpm::Log::instance()->setLogReportingLevel(antpm::LOG_DBG);
#endif

  // Declare the supported options.
  bool     pairing        = false;
  bool     dirOnly        = false;
  uint16_t dlFileIdx      = 0x0000;
  uint16_t eraseFileIdx   = 0x0000;
  int      verbosityLevel = antpm::Log::instance()->getLogReportingLevel();
  po::options_description desc("Allowed options");
  std::vector<const char*> args(argv, argv+argc);
  po::variables_map vm;
  try
  {
    desc.add_options()
    ("help,h",                                                                    "produce help message")
    ("pairing,P", po::value<bool>(&pairing)->zero_tokens()->implicit_value(true), "Force pairing first")
    ("dir-only",  po::value<bool>(&dirOnly)->zero_tokens()->implicit_value(true), "Download and list device directory")
    ("download,D",po::value<std::string>(),                                       "Download a single file (hex id e.g. 0x12FB) from device")
    ("erase",     po::value<std::string>(),                                       "Erase a single file (hex id e.g. 0x12FB) from device")
    ("v",                po::value<int>(&verbosityLevel),       "Adjust verbosity level, varies in [1, 6]")
    ("version,V",                                               "Print version information")
    ;

    //po::parsed_options parsed = po::parse_command_line(argc, argv, desc);
    po::parsed_options parsed = po::command_line_parser(argc, argv).options(desc).run();
    po::store(parsed, vm);

    if(vm.find("download")!=vm.end())
    {
      std::stringstream interpreter;
      interpreter << std::hex << vm["download"].as<std::string>();
      interpreter >> dlFileIdx;
    }
    if(vm.find("erase")!=vm.end())
    {
      std::stringstream interpreter;
      interpreter << std::hex << vm["erase"].as<std::string>();
      interpreter >> eraseFileIdx;
    }

    po::notify(vm);
  }
  catch(po::error& error)
  {
    cerr << error.what() << "\n";
    cerr << desc << "\n";
    return EXIT_FAILURE;
  }
  catch(std::exception& ex)
  {
    cerr << ex.what() << "\n";
    cerr << desc << "\n";
    return EXIT_FAILURE;
  }

  if(vm.count("version") || vm.count("V"))
  {
    cout << argv[0] << " " << antpm::getVersionString() << "\n";
    return EXIT_SUCCESS;
  }

  if(vm.count("v"))
  {
    if(verbosityLevel >= antpm::LOG_ERR && verbosityLevel <= antpm::LOG_DBG3)
    {
      antpm::Log::instance()->setLogReportingLevel(static_cast<antpm::LogLevel>(verbosityLevel));
    }
  }

  if(vm.count("help"))
  {
    logger() << desc << "\n";
    return EXIT_SUCCESS;
  }
  LOG_VAR5(pairing, dirOnly, dlFileIdx, eraseFileIdx, verbosityLevel);

  LOG(antpm::LOG_DBG) << argv[0] << " " << antpm::getVersionString() << "\n";
  for(int i = 0; i < argc; i++)
  {
    LOG(antpm::LOG_DBG2) << "\targv[" << i << "]\t\"" << argv[i] << "\"" << endl;
  }

  {
    AntFr310XT watch2;
    stopFunc = std::bind(&AntFr310XT::stopAsync, &watch2);
    {
      watch2.setModeDownloadAll();
      if(pairing) watch2.setModeForcePairing();
      if(dirOnly) watch2.setModeDirectoryListing();
      else if(dlFileIdx!=0x0000) watch2.setModeDownloadSingleFile(dlFileIdx);
      else if(eraseFileIdx!=0x000) watch2.setModeEraseSingleFile(eraseFileIdx);

      watch2.run();


      //watch2.stop();
    }
  }



  return EXIT_SUCCESS;
}
